WalletWasabi/Models/Label.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace WalletWasabi.Models
+{
+	public class Label : IEquatable<Label>
+	{
+		public static Label Empty { get; } = new Label();
+		public static char[] Separators { get; } = new[] { ',', ':' };
+		public IEnumerable<string> Labels { get; }
+		public bool IsEmpty { get; }
+
+		private string LabelString { get; }
+
+		public Label(params string[] labels) : this(labels as IEnumerable<string>)
+		{
+		}
+
+		public Label(IEnumerable<string> labels)
+		{
+			Labels = (labels
+				   ?.SelectMany(x => x?.Split(Separators, StringSplitOptions.RemoveEmptyEntries) ?? new string[0])
+				   ?.Select(x => x.Trim())
+				   ?.Where(x => x != "")
+				   ?.Distinct(StringComparer.OrdinalIgnoreCase)
+				   ?.OrderBy(x => x)
+				   ?? Enumerable.Empty<string>())
+				   .ToArray();
+
+			HashCode = ((IStructuralEquatable)Labels).GetHashCode(EqualityComparer<string>.Default);
+			IsEmpty = !Labels.Any();
+
+			LabelString = string.Join(", ", Labels);
+		}
+
+		public override string ToString() => LabelString;
+
+		public static Label Merge(IEnumerable<Label> labels)
+		{
+			return new Label(labels?
+				.SelectMany(x => x?.Labels ?? Enumerable.Empty<string>())
+				.Where(x => x != null)
+				?? Enumerable.Empty<string>());
+		}
+
+		public static Label Merge(params Label[] labels) => Merge(labels as IEnumerable<Label>);
+
+		#region Equality
+
+		public override bool Equals(object obj) => obj is Label label && this == label;
+
+		public bool Equals(Label other) => this == other;
+
+		private int HashCode { get; }
+
+		public override int GetHashCode() => HashCode;
+
+		public static bool operator ==(Label x, Label y)
+		{
+			if (x is null)
```

Comment:
I would go to readability here: 

```C#
			if (x is null && y is null)
			{
				return true;
			}

			if (x is null && y != null)
			{
				return false;
			}

			if (x != null && y == null)
			{
				return false;
			}

			return x.Labels.SequenceEqual(y.Labels);
```

Reply:
IMO Nopara's code is readable and more efficient.
If we check `if (x is null && y is null)` then `if (x is null && y != null)` ... there are some redundant checks.

Reply:
About readability, maybe it is a personal preference but I don't like if else inside another if else. For me it is too nested. I prefer to decrease nesting if possible. Let @lontivero decide about this.

Your second concern is not true according to my test:
https://dotnetfiddle.net/HTD4Bk

Reply:
I saw the same, to me nesting in this way is simply too much. I didn't request the change because it is not important enough. I would prefer something more flat. Personally I would go with somethin like this:

```c#
if ((x is null) && (y is null)) return true;
if ((x is null) != (y is null)) return false;
return x.Labels.SequenceEqual(y.Labels)
``

IMO this is concise and readable.

Reply:
This is a performance bottleneck. If anything, it should be optimized further. (The large label comparisions for clustering is calling this.)

Reply:
> Your second concern is not true according to my test:
https://dotnetfiddle.net/HTD4Bk

@molnard the results are `97905` and `1357572` if you pay attention the difference is huge in favor of your method, so clearly something is wrong.
In the first `for` loop where you used `MyCompare` method, this line `sw.Stop();` is inside the for loop so it stops in the first iteration. If you correct that and put it outside the for loop the result will be in favor of nopara's code. But the difference is not much anyway.

Reply:
There's the bug @yahiheb noted about the stop, but even correcting that David's compare is faster when I run it. Except if I switch the order, if my compare runs first, then it'll be the faster one. Which means this benchmark code is useless. Also null comparisons shouldn't be in the benchmark, our performance bottleneck is when we compare not nulls.

Reply:
Absolutely, my bad it was a copy-paste error - I also played with the order like nopara - but in VS the code was good and had a similar result which: there is no significant difference. 
I compared the edge cases because the rest of the algorithm is the same.

Reply:
Fixed: https://dotnetfiddle.net/HTD4Bk
Updated with Lucas' code.
The benchmark is random but it shows that the two comparisons are pretty the same regarding performance.

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/2201/commits/d854f712b57fab8bcc506525a15c5d385f75626a
https://github.com/zkSNACKs/WalletWasabi/pull/2201#issuecomment-528202575